### PR TITLE
Save pointer to store instead of object itself

### PR DESCRIPTION
**Pull request by TaoZou1**
Save pointer to store instead of object itself
    
 Avoid copy object into store, just save the pointer of object

### CODE REVIEW

File: pkg/nsx/services/ippool/store.go [hunk 1]
```diff
@@ -25,10 +25,14 @@ func keyFunc(obj interface{}) (string, error) {
 func indexFunc(obj interface{}) ([]string, error) {
 	res := make([]string, 0, 5)
 	switch v := obj.(type) {
-	case model.IpAddressPoolBlockSubnet:
+	case *model.IpAddressPoolBlockSubnet:
+		return filterTag(v.Tags), nil
+	case *model.IpAddressPool:
 		return filterTag(v.Tags), nil
 	case model.IpAddressPool:
```
Comment (zhengxiexie) [inline pkg/nsx/services/ippool/store.go:32]: shoud removed？

File: pkg/controllers/subnetset/vpc_handler.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		return false
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		return true
+	},
+	GenericFunc: func(genericEvent event.GenericEvent) bool {
+		return false
+	},
+}
```
Comment (heypnus) [inline pkg/controllers/subnetset/vpc_handler.go:118]: This file looks not related to the patch's topic...?

File: pkg/nsx/services/securitypolicy/firewall.go [hunk 3]
```diff
@@ -100,10 +100,18 @@ func (service *SecurityPolicyService) CreateOrUpdateSecurityPolicy(obj *v1alpha1
 	}
 
 	existingSecurityPolicy := service.securityPolicyStore.GetByKey(*nsxSecurityPolicy.Id)
-	existingRules := service.ruleStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
-	existingGroups := service.groupStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
-
+	tmpExistingRules := service.ruleStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
+	tmpExistingGroups := service.groupStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
+	existingRules := []model.Rule{}
```
Comment (heypnus) [inline pkg/nsx/services/securitypolicy/firewall.go:105]: L103~L111 looks a little strange... Can we directly change `func RulesToComparable(rules []model.Rule) []Comparable {...}` to `func RulesToComparable(rules []*model.Rule) []Comparable {...}` then directly pass to list of pointers to the function `RulesToComparable`? Then we don't need to convert the list of pointers to the list of objects before passing the list of objects to `RulesToComparable`...

File: pkg/nsx/services/securitypolicy/firewall.go [hunk 4]
```diff
@@ -100,10 +100,18 @@ func (service *SecurityPolicyService) CreateOrUpdateSecurityPolicy(obj *v1alpha1
 	}
 
 	existingSecurityPolicy := service.securityPolicyStore.GetByKey(*nsxSecurityPolicy.Id)
-	existingRules := service.ruleStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
-	existingGroups := service.groupStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
-
+	tmpExistingRules := service.ruleStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
```
Comment (timdengyun) [inline pkg/nsx/services/securitypolicy/firewall.go:103]: We can also change RulesToComparable, ComparableToRule, GroupsToComparable, ComparableToGroups to use pointer as well. So, we  don't need to use temporary rule pointer list here and convert them to teh ruls. We can directly use rule pointers in the compare related to function,  as we did for SecurityPolicyToComparable.

Comment (TaoZou1) [reply pkg/nsx/services/securitypolicy/firewall.go:103]: Done

File: pkg/nsx/services/securitypolicy/firewall.go (hunk 3 above)
Comment (TaoZou1) [reply pkg/nsx/services/securitypolicy/firewall.go:105]: Done

File: pkg/controllers/subnetset/vpc_handler.go (hunk 2 above)
Comment (TaoZou1) [reply pkg/controllers/subnetset/vpc_handler.go]: you're right. it may be introduced by wrong merge operation

**Comment by timdengyun** [comment]
do we still need this PR https://github.com/vmware-tanzu/nsx-operator/pull/365?

File: pkg/nsx/services/securitypolicy/compare.go [hunk 5]
```diff
@@ -89,18 +89,32 @@ func SecurityPolicyToComparable(sp *model.SecurityPolicy) Comparable {
 	return (*SecurityPolicy)(sp)
 }
 
-func RulesToComparable(rules []model.Rule) []Comparable {
+func RulesPtrToComparable(rules []*model.Rule) []Comparable {
```
Comment (timdengyun) [inline pkg/nsx/services/securitypolicy/compare.go:92]: why not change SecurityPolicyToComparable to SecurityPolicyPtrToComparable to make naming convention  consistent with RulesPtrToComparable, GroupsPtrToComparable etc.

**Review by heypnus — APPROVED** [review]

File: pkg/nsx/services/securitypolicy/store.go [hunk 6]
```diff
@@ -114,35 +114,36 @@ func (securityPolicyStore *SecurityPolicyStore) Apply(i interface{}) error {
...
-		securityPolicies = append(securityPolicies, securityPolicy.(model.SecurityPolicy))
+		securityPolicies = append(securityPolicies, securityPolicy.(*model.SecurityPolicy))
 	}
 	return securityPolicies
 }
 
 func (ruleStore *RuleStore) Apply(i interface{}) error {
 	sp := i.(*model.SecurityPolicy)
 	for _, rule := range sp.Rules {
+		tempRule := rule
```
Comment (timdengyun) [inline pkg/nsx/services/securitypolicy/store.go:137]: why we always need a temp var: tempRule when iterating sp.rules? 
we see this is happening for other Apply functions as well.

Comment (TaoZou1) [reply pkg/nsx/services/securitypolicy/store.go:137]: to remove warning by lint

Comment (timdengyun) [reply pkg/nsx/services/securitypolicy/store.go:137]: what lint do you use? golang-ci-lint?
